keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md

```diff
@@ -847,12 +981,24 @@ Longer term, we may want to require automated upgrade/rollback tests, but we
 are missing a bunch of machinery and tooling and can't do that now.
 -->
 
+For alpha, manual tests were performed to:
+
+- Unsuccessfully create workloads using ReadWriteOncePod PVCs prior to upgrade
+- Perform the upgrade (enabling feature flags and updating CSI drivers)
+- Successfully create workloads using ReadWriteOncePod PVCs
+- Perform the downgrade (disabling feature flags and downgrading CSI drivers)
+- Successfully delete ReadWriteOncePod PVCs
```

Comment:
Revised this to include steps for the additional re-upgrade after downgrading. CSI functionality is being exercised here during the provisioning of PVCs (calling `CreateVolume` for provisioning, `ControllerPublishVolume` for attachment if supported, `NodePublishVolume` for mounting, etc.).

Two things that stood out while going through this exercise:
- The CSI sidecars cannot be downgraded if one wants to delete ReadWriteOncePod PVCs, otherwise it will stall "Terminating" because the CSI sidecar does not understand this access mode in older versions
- If a user disables the ReadWriteOncePod feature, they *can* create a pod that mounts an existing ReadWriteOncePod PVC, but they cannot create a new ReadWriteOncePod PVC
  - I'm thinking this is expected behavior since the feature is disabled (CC @msau42)
  - Our validation logic allows us to continue working with existing ReadWriteOncePod PVCs, this is just a byproduct

Reply:
Re: sidecars cannot be downgraded
I don't see Delete() blocking on access mode: https://github.com/kubernetes-csi/external-provisioner/blob/dfaeafbe23eec030cd0a6146aa24bb482fd3f918/pkg/controller/controller.go#L1197

nor detach: https://github.com/kubernetes-csi/external-attacher/blob/553a354f0bd5362e8c7523750ccdd448df4509b4/pkg/controller/csi_handler.go#L420

So I think deleting PVCs should work.

Reply:
Re: using the PVC after feature is disabled

It looks like both the scheduler and kubelet enforcements are feature gated, so someone can still use the PVC, however none of the "one pod" enforcement will be active:
https://github.com/kubernetes/kubernetes/pull/103082/files#diff-43b8bd344b43e1bc9011cd28f976d70b321abddbbb95bd958d0b5222305b1da4R124

https://github.com/kubernetes/kubernetes/pull/102028/commits/2b98f8edc77455f3c23a3156f2c56ffa9e643b85#diff-3e8f8a6e5eca57db8e84bcc6c896f0a3b10289fe55d8e999b4bdf1cfdc0de394R547

Reply:
Oh, thanks for pointing this out. I think that was a byproduct of me using the mock CSI driver to test. Because it stores volumes in memory, when I downgraded it it restarted and so it failed to delete because it didn't know about those volumes anymore.